packages/plugins/apollo-usage-report/src/reporter.ts

```diff
@@ -0,0 +1,186 @@
+    const report = this.#getReport(schemaId);
+    report.addTrace(options);
+
+    if (
+      this.#options.alwaysSend ||
+      report.sizeEstimator.bytes >= this.#options.maxBatchUncompressedSize!
+    ) {
+      return this.sendReport(schemaId);
+    }
+
+    this.#nextSendAfterDelay ||= setTimeout(() => this.flush(), this.#options.maxBatchDelay);
+
+    return;
+  }
+
+  async flush() {
+    return Promise.allSettled([
+      ...this.#sending, // When flushing, we want to also wait for previous traces to be sent, because it's mostly used for clean up
+      ...Object.keys(this.#reportsBySchema).map(schemaId => this.sendReport(schemaId)),
+    ]);
+  }
+
+  async sendReport(schemaId: string) {
+    const sending = this.#sendReport(schemaId);
+    this.#sending.push(sending);
+    sending.finally(() => this.#sending?.filter(p => p !== sending));
+    return sending;
+  }
+
+  async #sendReport(schemaId: string) {
+    const {
+      fetchAPI: { fetch, CompressionStream, ReadableStream },
+    } = this.#yoga;
+    const report = this.#reportsBySchema[schemaId];
+    if (!report) {
+      throw new Error(`No report to send for schema ${schemaId}`);
+    }
+
+    if (this.#nextSendAfterDelay != null) {
+      clearTimeout(this.#nextSendAfterDelay);
+      this.#nextSendAfterDelay = undefined;
+    }
+
+    delete this.#reportsBySchema[schemaId];
+    report.endTime = dateToProtoTimestamp(new Date());
+    report.ensureCountsAreIntegers();
+
+    const validationError = Report.verify(report);
+    if (validationError) {
+      throw new TypeError(`Invalid report: ${validationError}`);
+    }
+
+    const { apiKey = getEnvVar('APOLLO_KEY'), endpoint = DEFAULT_REPORTING_ENDPOINT } =
+      this.#options;
+
+    const encodedReport = Report.encode(report).finish();
+
+    for (let tries = 0; tries < 5; tries++) {
+      try {
+        this.#logger?.debug(`Sending report (try ${tries}/5)`);
+        const response = await fetch(endpoint, {
+          method: 'POST',
+          headers: {
+            'content-type': 'application/protobuf',
+            'content-encoding': 'gzip',
+            // The presence of the api key is already checked at Yoga initialization time
+            'x-api-key': apiKey!,
+            accept: 'application/json',
+          },
+          body: new ReadableStream({
+            start(controller) {
+              controller.enqueue(encodedReport);
+              controller.close();
+            },
+          }).pipeThrough(new CompressionStream('gzip')),
+          signal: AbortSignal.timeout(this.#options.exportTimeout),
+        });
+
+        const result = await response.text();
+        if (response.ok) {
+          this.#logger?.debug('Report sent:', result);
+          return;
+        }
+
+        throw result;
+      } catch (err) {
+        this.#logger?.error('Failed to send report:', err);
+      }
+    }
+
+    throw new Error('Failed to send traces after 5 tries');
```

Comment:
@EmrysMyrddin This error has already popped up once in our NestJS microservice as "unhandled error". I would rather not throw an error here, but just log this state

Reply:
But now that I think about it, it probably doesn't matter. It's an "Unhandled error" because it's sent inside the `setTimeout()` method and it shouldn't break incoming requests if the endpoint `usage-reporting.api.apollographql.com/api/ingress/traces` is unavailable

Reply:
Yes, that would probably the be best. Something like that:
```
this.sendReport(schemaId).catch((error) => {
  if (options.onError) {
    options.onError(error);
  } else {
   this.#logger.error(error);
  }
```